### DISCUSSION TITLE
Apple Pay is not supported on a physical iPhone device.

### DISCUSSION THREAD

**Discussion by SebghatYusuf** [flutter-stripe/flutter_stripe#690] [Q&A] [answered]
**Describe the bug**
I'm using the in-package ApplePay and Google Pay integration to integrate Apple Pay & Gpay with my app. GPay works fine on real devices, but the main issue is with Apple Pay, I'm able to test and pay (Generate Stripe Token and complete the payment) on Simulator. 
When I ran the app on my iPhone device in debug mode `await Stripe.instance.checkApplePaySupport();` always returns `false` even when I have valid/active cards added in my Apple Wallet.

I've followed this guide [Apple Pay | Stripe Documentation](https://stripe.com/docs/apple-pay) step by step as follows to setup the apple pay: 

1- [Register for an Apple Merchant ID](https://stripe.com/docs/apple-pay?platform=ios#merchantid)
![Merchant ID](https://user-images.githubusercontent.com/48417629/164403418-9b7f4758-70a6-40ae-a6da-64ccd3c6aa10.png)

2- [Create a new Apple Pay certificate](https://stripe.com/docs/apple-pay?platform=ios#csr)
![Stripe - Dashboard certificate](https://user-images.githubusercontent.com/48417629/164403473-157ffec9-e644-4fb5-8209-838b4b225ef2.png)
![Apple - Stripe merchant certificate](https://user-images.githubusercontent.com/48417629/164403512-4a728aed-d13a-446c-a802-ece9c4b3b264.png)

3- [Integrate with Xcode](https://stripe.com/docs/apple-pay?platform=ios#xcode-pay)
![xcode](https://user-images.githubusercontent.com/48417629/164403562-3c474d3e-775f-4283-b120-3a810411735b.png)

The 4th (5 in docs) step is failing : 

4- [Check if Apple Pay is supported](https://stripe.com/docs/apple-pay?platform=ios#check-if-apple-pay-supported)

here is code snippet from checkout page `initState`

```
Stripe.publishableKey = pubKey;
Stripe.merchantIdentifier = "merchant ID as created in Apple merchant ID and set in xcode";
await Stripe.instance.applySettings();
```

**To Reproduce**
1. create a fresh project. 
2. add flutter_stripe and setup the apple pay.
3. test on a simulator and then run on a physical device (iPhone)

**Expected behavior**
I should be able to run my app in debug mode on my iPhone and be able to see the Payment Sheet as I can see in the simulator. 
![Sim 1](https://user-images.githubusercontent.com/48417629/164403270-cd0d838b-4a9c-45c8-aa69-a43e3f31cb61.png)
![Sim 2](https://user-images.githubusercontent.com/48417629/164403329-31e9c1b0-efb7-4f37-af75-cb21663630aa.png)

**Smartphone / tablet**
 - Device: [iPhone 7,  iPhone 14]
 - OS: [iOS 15.4.1, iOS 15.0]
 - Package version: 2.4.0
 - Flutter version 2.10.1

**Additional context**
Please let me know what step I'm not doing correctly if the issue is from my side.

**Comment by remonh87** [comment]
@SebghatYusuf  can you share the initpaymentsheet call ?  can be sth is missing in the paymentsheet parameters

**Reply by SebghatYusuf** [reply]
@remonh87 
I'm not invoking any initPaymentSheet method/function, basically here is the code I have 
for the `initState` and initializing Stripe I've shared the code snippet above. but after that I'm doing the following steps: 

1. Adding a widget to my checkout page widget tree: 

```
if (supportApplePay)
 Expanded(
  child: SizedBox(
    height: 40,
    width: double.infinity,
    child: stripe.ApplePayButton(
      onPressed: _handleApplePayPress,
      style: stripe
          .ApplePayButtonStyle
          .automatic,
    ),
  ),
),

```


2. and when user taps on the button here's the `_handleApplePayPress` function : 

```

Future<void> _handleApplePayPress() async {
    try {
      var support = await stripe.Stripe.instance.checkApplePaySupport();
      // print('supprot $support');
      if (support) {
        // 1. Present Apple Pay sheet
        await stripe.Stripe.instance.presentApplePay(
          stripe.ApplePayPresentParams(
            cartItems: [
              stripe.ApplePayCartSummaryItem(
                label: widget.plan['name'] ?? 'Test Plan',
                amount: widget.plan['price']?.toString() ?? "0.01",
              ),
            ],
            country: 'us',
            currency: "usd",
          ),
        );

        // 2. fetch Intent Client Secret from backend
        final response = await provider!.fetchPaymentIntentClientSecret(
          currency: 'usd',
          plan_name: widget.plan['name'],
          price: widget.plan['price'].toString(),
          tracking_id: trackingId,
        );
        if (response['status'] != 'success') return;
        var data = response['data'];
        final clientSecret = data['client_secret'];

        // 2. Confirm apple pay payment
        await stripe.Stripe.instance.confirmApplePayPayment(clientSecret);
        setState(() {
          _loaderTitle = 'Processing...';
        });
        _onPaymentSuccess();
      } else {
        ScaffoldMessenger.of(context).showSnackBar(
          const SnackBar(
              content: Text('Apple pay is not supported on your device.')),
        );
      }
    } catch (e) {
      print(e);
      const SnackBar(
        content: Text('Apple pay is not supported on your device.'),
      );
    }
  }

```

**Reply by SebghatYusuf** [reply]
@remonh87 Any updates?

**Reply by SebghatYusuf** [reply]
Isn't there any solution for using the Apple Pay with Stripe in Flutter? 
Or the steps I followed are incorrect, 
Any inputs are welcome. 
Thanks in advance.

**Reply by remonh87** [reply]
Did you create a merchant id on Apple and configured it in the stripe settings + you also need to add the key to the stripe dashboard. See step 2 in the docs  https://stripe.com/docs/payments/accept-a-payment?platform=ios . Your code looks good and is the same as mine

**Reply by SebghatYusuf** [reply]
@remonh87 
Yes, I did create a merchant ID on Apple and configured it in the stripe settings, as the description above shows in steps 1 and 2. 
Also, I have enabled Apple and Google Pay, as payment methods (step 2 in the docs). 
![Screen Shot 2022-05-27 at 2 50 20 PM](https://user-images.githubusercontent.com/48417629/170681470-f65b267b-4020-42b6-8dca-8133e5228bf2.png)

**Reply by SlickSlime** [reply]
@SebghatYusuf did you get any resolution or workaround for this issue? I am facing the same issue. It's working in the iOS Simulator but not on physical device.

**Reply by SebghatYusuf** [reply]
@SlickSlime Unfortunately no, as we're releasing only for Android, for now, I haven't looked into the issue yet, I believe it's a regional issue, I tried to have the same merchant identifier as the app bundle but that also did not work.

**Reply by SlickSlime** [reply]
"regional issue" meaning the reason is because Apple Pay is not available in your (and in my case India) region? @SebghatYusuf

**Reply by SebghatYusuf** [reply]
@SlickSlime 
Yes, It might be, but still, I'm not sure.

**Comment by SlickSlime** [comment, marked as answer]
From spending a day on this issue and chatting with Stripe's Customer Support team.
- [ ] Check this link (https://stripe.com/docs/stripe-js/elements/payment-request-button). You'll see the Apple Pay button at the top of the page if everything is as it should be.
- [ ] You need to have at least 1 card in your Apple Pay wallet for `await Stripe.instance.checkApplePaySupport();` to return `true`.

In iOS Simulator, few cards are already saved in the wallet, so Apple Pay was "available" on the Simulator. In my case, the reason Apple Pay was "unavailable" on the physical device was that I am in India (where Apple Pay is not available) + there were no cards available in the wallet. 

![image](https://user-images.githubusercontent.com/19946756/174829359-8d83fbce-8c40-4982-9f83-6bbe78e7e97c.png)

**Reply by SebghatYusuf** [reply]
@SlickSlime 
Amazing, this is great! 
thanks for sharing. actually, I've tried to test the app on a device with apple cards available in the wallet and it was in Canada (I asked them to test), still the `await Stripe.instance.checkApplePaySupport();` was returning `false` but that time my merchant identifier was not same as bundle identifier. I will check this too
Please let me know if it works well for you. :)

**Reply by SlickSlime** [reply]
I doubt the issue would be related to the "sameness" of the merchant identifier as in our implementation it doesn't match but still works for users with a card in their Apple Wallet. But 🤞🏽.

PS You might want to chat with Stripe's Customer Support, they are pretty helpful.

**Comment by simonpham** [comment]
So bad that Apple is rejecting the app because of this.

> We continue to see the same bug in your app during our review on real iPad devices running iOS 15.4.1, where the Apple Pay button still does not come up after choosing Apple Pay as a payment method. Please note, that when Apple Pay is not set up on a device, the app should direct users to set up Apple Pay.
